k2/python/k2/ops.py

```diff
@@ -68,3 +122,83 @@ def index_add(index: torch.Tensor, value: torch.Tensor,
     '''
 
     _k2.index_add(index, value, in_out)
+
+
+def index_ragged_int(src: _k2.RaggedInt,
+                     indexes: Union[torch.Tensor, _k2.RaggedInt]
+                    ) -> _k2.RaggedInt:
+    '''Indexing ragged tensor with a 1-D tensor or a ragged tensor.
+
+    Args:
+      src:
+        Source ragged tensor to index; must have num_axes() == 2.
+      indexes:
+        It can be a tensor or a ragged tensor.
+        If it's a tensor, it must be a 1-D tensor and `indexes.dtype == torch.int32`.
+        Values in it will be interpreted as indexes into axis 0 of `src`, 
+        i.e. 0 <= indexes[i] < src.dim0().
+        If it's a ragged tensor, `indexes.values` will be interpreted as indexes into 
+        axis 0 of `src`, i.e. 0 <= indexes.values[i] < src.dim0(); Must have 
+        num_axes() == 2.
+
+    Returns:
+      Return the indexed ragged tensor with ans.num_axes() == 2
+       - If `indexes` is a 1-D tensor, then ans.dim0() == indexes.numel().
+       - If `indexes` is a ragged tensor, then ans.dim0() = indexes.dim0().
+    '''
+    if isinstance(indexes, torch.Tensor):
+        ans, _ = ragged_index(src, indexes)
+        return ans
+    else:
+        return _k2.index_ragged_with_ragged_int(src, indexes)
+
+
+def index_tensor_with_ragged_int(src: torch.Tensor,
+                                 indexes: _k2.RaggedInt) -> _k2.RaggedInt:
+    '''Indexing a 1-D tensor with a ragged tensor.
+
+    Args:
+      src:
+        Source 1-D tensor to index, must have `src.dtype == torch.int32`
+      indexes:
+        A ragged tensor, `indexes.values` will be interpreted as indexes into `src`.
+        i.e. 0 <= indexes.values[i] < src.numel();
+
+    Returns:
+      Returns ragged tensor with shape `indexes.shape` and values `src[indexes.values]`.
+    '''
+    return _k2.index_tensor_with_ragged_int(src, indexes)
+
+
+def index_tensor(src: torch.Tensor, indexes: Union[torch.Tensor, _k2.RaggedInt]
+                ) -> Union[torch.Tensor, _k2.RaggedInt]:
+    '''It's a wrapper of index_tensor and index_ragged_int above
+
+    Args:
+      src:
+        Source 1-D tensor to index, must have `src.dtype == torch.int32`
+      indexes:
+        If it's a ragged tensor, `indexes.values` will be interpreted as indexes into `src`.
+        i.e. 0 <= indexes.values[i] < src.numel(); 
+        If it's a tensor, its values will be interpreted as indexes into `src`.
+
+    Returns:
+      Returns a tensor or a ragged tensor (depending on the type of `indexes`)
+    '''
+    if isinstance(indexes, torch.Tensor):
+        return index_select(src, indexes)
+    else:
+        # TODO(haowen): it does not autograd now.
+        return index_tensor_with_ragged_int(src, indexes)
+
+
+def index_attr(src: Union[torch.Tensor, _k2.RaggedInt],
```

Comment:
can you explain the name of this?

Reply:
maybe not a good name, I called it `index_attribute(of fsa)`, but that is a general index operation, any suggestion?